### PR TITLE
Accept bicep file, convert it to JSON, analyze the JSON

### PR DESCRIPTION
**Pull request by norbusonam**
* Currently, we only show errors in the JSON, and do not relate it back to the bicep

Updates (nonik0):
* using published version of Azure.Bicep.Core
* using new source map from Azure.Bicep.Core to translate errors back to original bicep

### CODE REVIEW

File: src/Analyzer.BicepProcessor/Analyzer.BicepProcessor.csproj [hunk 1]
```diff
@@ -0,0 +1,27 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (yane3628) [inline src/Analyzer.BicepProcessor/Analyzer.BicepProcessor.csproj:1]: Can you enforce summary comments on public functions. Here is an example of how to enable that in the csproj https://github.com/Azure/template-analyzer/blob/c9f38aa454ced1d6d689bee9d9c37fdfe2ada826/src/Analyzer.JsonRuleEngine/Analyzer.JsonRuleEngine.csproj#L10

https://github.com/Azure/template-analyzer/blob/c9f38aa454ced1d6d689bee9d9c37fdfe2ada826/src/Analyzer.JsonRuleEngine/Analyzer.JsonRuleEngine.csproj#L11

File: src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
+﻿using Microsoft.Azure.Templates.Analyzer.TemplateProcessor;
```
Comment (yane3628) [inline src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs:1]: Add copyright header to the top of file

``` c#
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
```

Comment (yane3628) [reply src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs:1]: Move the System using statement to the top of the using statements and then alphabetize the remaining below it.

File: src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    public class BicepTemplateProcessor
+    {
+
+        private readonly string bicep;
+        private readonly string bicepPath;
+        private readonly string apiVersion;
+        private readonly bool dropResourceCopies;
+
+        /// <summary>
+        ///  Constructor for the ARM Template Processing library
```
Comment (yane3628) [inline src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs:21]: Comments need to be updated to reflect bicep

File: src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+using Bicep.Core.Semantics;
+using Bicep.Core.Syntax;
+using Bicep.Core.Workspaces;
+using Bicep.Core.TypeSystem.Az;
+
+namespace Microsoft.Azure.Templates.Analyzer.BicepProcessor
+{
+    public class BicepTemplateProcessor
+    {
+
```
Comment (yane3628) [inline src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs:14]: Minor: Remove new line

File: src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        /// <param name="apiVersion">The deployment API version. Must be a valid version from the deploymetns list here: https://docs.microsoft.com/en-us/azure/templates/microsoft.resources/allversions</param>
+        /// <param name="dropResourceCopies">Whether copies of resources (when using the copy element in the ARM Template) should be dropped after processing.</param>
+        public BicepTemplateProcessor(string bicepTemplate, string bicepPath, string apiVersion = "2020-01-01", bool dropResourceCopies = false)
+        {
+            this.bicep = bicepTemplate;
+            this.bicepPath = bicepPath;
+            this.apiVersion = apiVersion;
+            this.dropResourceCopies = dropResourceCopies;
+        }
+        public string ConvertBicepToJson()
```
Comment (yane3628) [inline src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs:34]: Minor: Add new line

File: src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        {
+            using var stringWriter = new StringWriter();
+            var syntaxTreeGrouping = SyntaxTreeGroupingBuilder.Build(new FileResolver(), new Workspace(), PathHelper.FilePathToFileUrl(bicepPath));
+            var compilation = new Compilation(AzResourceTypeProvider.CreateWithAzTypes(), syntaxTreeGrouping);
+            var emitter = new TemplateEmitter(compilation.GetEntrypointSemanticModel(), "0.4.123.62267");
```
Comment (yane3628) [inline src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs:39]: Should this assembly file version be able to be passed through the constructor or will it never change? Is it defined as an enum or something in the Bicep library?

File: src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs (hunk 2 above)
Comment (yane3628) [reply src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs:1]: `Microsoft.Azure.Templates.Analyzer.TemplateProcessor` is not used so you can remove it

File: src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+using Bicep.Core.Emit;
+using Bicep.Core.FileSystem;
+using Bicep.Core.Semantics;
+using Bicep.Core.Syntax;
+using Bicep.Core.Workspaces;
+using Bicep.Core.TypeSystem.Az;
+
+namespace Microsoft.Azure.Templates.Analyzer.BicepProcessor
+{
+    public class BicepTemplateProcessor
```
Comment (yane3628) [inline src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs:12]: Add summary comment for this class

File: src/Analyzer.Core/TemplateAnalyzer.cs [hunk 8]
```diff
@@ -35,6 +37,7 @@ public TemplateAnalyzer(string template, string parameters = null, string templa
             this.Template = template ?? throw new ArgumentNullException(paramName: nameof(template));
             this.Parameters = parameters;
             this.TemplateFilePath = templateFilePath;
+            this.IsBicep = templateFilePath != null && templateFilePath.ToLower().EndsWith(".bicep");
```
Comment (yane3628) [inline src/Analyzer.Core/TemplateAnalyzer.cs:40]: Add a test for the updates made to this constructor.

File: src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        private readonly bool dropResourceCopies;
+
+        /// <summary>
+        ///  Constructor for the ARM Template Processing library
+        /// </summary>
+        /// <param name="bicepTemplate">The ARM Template <c>JSON</c>. Must follow this schema: https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#</param>
+        /// <param name="bicepPath">The ARM Template <c>JSON</c>. Must follow this schema: https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#</param>
+        /// <param name="apiVersion">The deployment API version. Must be a valid version from the deploymetns list here: https://docs.microsoft.com/en-us/azure/templates/microsoft.resources/allversions</param>
+        /// <param name="dropResourceCopies">Whether copies of resources (when using the copy element in the ARM Template) should be dropped after processing.</param>
+        public BicepTemplateProcessor(string bicepTemplate, string bicepPath, string apiVersion = "2020-01-01", bool dropResourceCopies = false)
```
Comment (yane3628) [inline src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs:27]: Unless we need them later, I believe the only thing this constructor/class needs is the file path. I recommend removing the rest and adding any in the future as needed.

File: src/TemplateAnalyzer.sln [hunk 10]
```diff
@@ -39,6 +39,8 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Analyzer.PowerShellRuleEngi
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Analyzer.PowerShellRuleEngine.UnitTests", "Analyzer.PowerShellRuleEngine.UnitTests\Analyzer.PowerShellRuleEngine.UnitTests.csproj", "{D6664E6D-827C-49BE-AD94-49977605E780}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Analyzer.BicepProcessor", "Analyzer.BicepProcessor\Analyzer.BicepProcessor.csproj", "{FC3CC274-6E47-40F7-9948-71C2DE09A0C8}"
```
Comment (yane3628) [inline src/TemplateAnalyzer.sln:42]: Can you make a solution folder and put this project in it?

**Review by yane3628 — CHANGES_REQUESTED** [review]

File: src/Analyzer.Core/TemplateAnalyzer.cs (hunk 8 above)
Comment (norbusonam) [reply src/Analyzer.Core/TemplateAnalyzer.cs:40]: _Discussed on teams_: don't need these tests for now.

File: src/Analyzer.BicepProcessor/Analyzer.BicepProcessor.csproj (hunk 1 above)
Comment (yane3628) [reply src/Analyzer.BicepProcessor/Analyzer.BicepProcessor.csproj:1]: @norbusonam There's a PR that generalizes this now https://github.com/Azure/template-analyzer/pull/151 You can remove the additional lines in the csproj.

File: src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs (hunk 6 above)
Comment (yane3628) [reply src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs:40]: @anthony-c-martin Does the `assemblyFileVersion` parameter for `TemplateEmitter` matter for our context? Can we set it to `null` or an empty string? I don't want to have to update this value in the future if it's not necessary.

Comment (anthony-c-martin) [reply src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs:40]: You can set it to empty string if you like - it's only used to stamp metadata into the generated JSON for telemetry purposes (example [here](https://github.com/Azure/bicep/blob/f1a4461a84bdb7b3c0f5e329b64387b0ba73fe72/docs/examples/101/1vm-2nics-2subnets-1vnet/main.json#L7)).

**Review by VeraBE — APPROVED** [review]

**Review by JohnathonMohr — CHANGES_REQUESTED** [review]

File: src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    /// </summary>
+    public class BicepTemplateProcessor
+    {
+        private readonly string bicepPath;
+
+        /// <summary>
+        ///  Constructor for the Bicep Template Processing library
+        /// </summary>
+        /// <param name="bicepPath">The Bicep Template file path. Needed to convert to JSON.</param>
+        public BicepTemplateProcessor(string bicepPath)
```
[the remark this answers is not in the document]
Comment (norbusonam) [reply src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs:26]: removed constructor and made ConvertBicepToJson a static function.

File: src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+                configuration,
+                new LinterAnalyzer(configuration));
+
+            var settings = new EmitterSettings(featureProvider);
+            var emitter = new TemplateEmitter(compilation.GetEntrypointSemanticModel(), settings);
+            var emitResult = emitter.Emit(stringWriter);
+
+            if (emitResult.Status == EmitStatus.Failed)
+            {
+                var bicepIssues = string.Join('\n', emitResult.Diagnostics.Select(diag => diag.Message));
```
Comment (anthony-c-martin) [inline src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs:62]: Technically you're also going to be displaying warnings & info diagnostics here which won't actually fail the build. It might be worth either filtering to just errors, or including the level as part of the message.

File: src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+        {
+            using var stringWriter = new StringWriter();
+
+            Environment.SetEnvironmentVariable("BICEP_SOURCEMAPPING_ENABLED", "true");
+
+            var fileResolver = new FileResolver();
+            var featureProvider = new FeatureProvider();
+            
+            var moduleDispatcher = new ModuleDispatcher(
+                new DefaultModuleRegistryProvider(
```
Comment (anthony-c-martin) [inline src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs:42]: I don't know how important the module registry is for the analyzer, but it might be worth verifying what the behavior is like if you attempt to access a .bicep file from a module registry or a template spec - e.g.
```bicep
module gridVnet 'br:Carml.azurecr.io/bicep/modules/microsoft.network.virtualnetworks:v0.4.0' = {
  name: 'foo'
}
```

Comment (nonik0) [reply src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs]: Yeah, as expected I suppose causes issues when building JSON:
`The module with reference &quot;br:carml.azurecr.io/bicep/modules/microsoft.network.virtualnetworks:v0.4.0&quot; has not been restored.`

Comment (nonik0) [reply src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs]: I don't see any other implementations for IModuleRegistryProvider, and I am also not sure how this would work with the source map either. Do you have a little more context here or thoughts about how we should be handling these sorts of bicep files?

Comment (anthony-c-martin) [reply src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs]: I think best you can do is handle the case nicely where authentication fails. I believe if you were to log in with az CLI or az powershell before building the Bicep file, and you have access to the registry, module restore should work, but there will definitely be cases where the registry is private and auth isn't possible.

Comment (anthony-c-martin) [reply src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs]: If you do want to be able to pull down modules optimistically, you should be able to use:
https://github.com/Azure/bicep/blob/d46c7d2ee2c78afc0dad767fa21f78b352a7d4d1/src/Bicep.Cli/Services/CompilationService.cs#L58-L64

File: src/Analyzer.BicepProcessor/BicepLocationResolver.cs [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+            // source map line numbers from bicep are 0-indexed
+            jsonLine--;
+
+            // TODO: look for mappings in other files/modules once nested templates are supported, for now just entrypoint file
+            var entrypointFile = this.sourceMap.Entries.First(entry => entry.FilePath == this.sourceMap.Entrypoint);
+            var match = entrypointFile.SourceMap.FirstOrDefault(mapping => mapping.TargetLine == jsonLine);
+
+            return (match != null)
+                ? match.SourceLine + 1 // convert to 1-indexing
+                : 0;
```
Comment (VeraBE) [inline src/Analyzer.BicepProcessor/BicepLocationResolver.cs:47]: I think we should return 1 when it can't be resolved so it's consistent with how the JsonLineNumberResolver behaves

File: src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs [hunk 15]
```diff
@@ -30,6 +30,8 @@ public void TestInit()
         [DataRow("Configuration.json", ExitCode.ErrorCommand, "--parameters-file-path", DisplayName = "Path exists, Parameters-file-path flag included, but no value provided.")]
         [DataRow("AppServicesLogs-Failures.json", ExitCode.Violation, DisplayName = "Violations found in the template")]
         [DataRow("AppServicesLogs-Passes.json", ExitCode.Success, DisplayName = "Success")]
+        [DataRow("AppServicesLogs-Failures.bicep", ExitCode.Violation, DisplayName = "Violations found in the template")]
```
Comment (VeraBE) [inline src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs:33]:
```suggestion
        [DataRow("AppServicesLogs-Failures.bicep", ExitCode.Violation, DisplayName = "Violations found in the Bicep template")]
```

File: src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs [hunk 16]
```diff
@@ -30,6 +30,8 @@ public void TestInit()
...
+        [DataRow("AppServicesLogs-Passes.bicep", ExitCode.Success, DisplayName = "Success")]
```
Comment (VeraBE) [inline src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs:34]:
```suggestion
        [DataRow("AppServicesLogs-Passes.bicep", ExitCode.Success, DisplayName = "Success with a Bicep template")]
```

File: src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs [hunk 17]
```diff
@@ -125,7 +127,7 @@ public void AnalyzeDirectory_DirectoryWithInvalidTemplates_LogsExpectedErrorInSa
             Assert.AreEqual(templateErrorMessage, toolNotifications[1]["message"]["text"]);
 
             var nonJsonFilePath1 = Path.Combine(directoryToAnalyze, "AnInvalidTemplate.json");
-            var nonJsonFilePath2 = Path.Combine(directoryToAnalyze, "AnotherInvalidTemplate.json");
```
Comment (VeraBE) [inline src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs:128]: Why did you remove this file? It was to test the code that shows multiple paths with errors

File: src/Analyzer.Utilities/ILocationResolver.cs [hunk 18]
```diff
@@ -6,8 +6,10 @@ namespace Microsoft.Azure.Templates.Analyzer.Utilities
     /// <summary>
     /// An interface used for resolving line numbers in a document.
     /// </summary>
-    public interface ILineNumberResolver
+    public interface ILocationResolver
     {
+        // TODO: include source file for bicep once nested templates are supported
```
Comment (VeraBE) [inline src/Analyzer.Utilities/ILocationResolver.cs:11]: Maybe we should add a comment mentioning this in the nested templates issue/ADO story?

File: src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs (hunk 17 above)
Comment (nonik0) [reply src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs:128]: My intention is that this file was replaced with AnInvalidTemplate.bicep , does this not provide the same coverage?

File: src/Analyzer.BicepProcessor/BicepLocationResolver.cs (hunk 14 above)
Comment (nonik0) [reply src/Analyzer.BicepProcessor/BicepLocationResolver.cs:47]: lol then I will also update the comment in JsonLineNumberResolver that made me go with 0 to begin with

File: src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs (hunk 13 above)
Comment (nonik0) [reply src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs]: OK added the restore functionality, though I wasn't able to auth to this registry still. Will look into soon as modules are not going to be initially supported until the nested template work in progress is closer to complete.

File: src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs (hunk 17 above)
Comment (VeraBE) [reply src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs:128]: Oh, you're right, never mind

**Review by VeraBE — APPROVED** [review]

File: src/Analyzer.BicepProcessor/BicepLocationResolver.cs [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+        /// </summary>
+        /// <param name="pathInExpandedTemplate">The path in the expanded template
+        /// to find the line number of in the original template.</param>
+        /// <returns>The line number of the equivalent location in the original template,
+        /// or 1 if it can't be determined.</returns>
+        public int ResolveLineNumber(string pathInExpandedTemplate)
+        {
+            var jsonLine = this.jsonLineNumberResolver.ResolveLineNumber(pathInExpandedTemplate);
+
+            // source map line numbers from bicep are 0-indexed
```
Comment (anaismiller) [inline src/Analyzer.BicepProcessor/BicepLocationResolver.cs:38]:
```suggestion
            // Source map line numbers from bicep are 0-indexed
```

File: src/Analyzer.BicepProcessor/BicepLocationResolver.cs [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+            jsonLine--;
+
+            // TODO: look for mappings in other files/modules once nested templates are supported, for now just entrypoint file
+            var entrypointFile = this.sourceMap.Entries.First(entry => entry.FilePath == this.sourceMap.Entrypoint);
```
Comment (anaismiller) [inline src/Analyzer.BicepProcessor/BicepLocationResolver.cs:42]: Can this be entryPointFile instead of entrypointFile?

File: src/Analyzer.BicepProcessor/BicepLocationResolver.cs (hunk 19 above)
Comment (anaismiller) [inline src/Analyzer.BicepProcessor/BicepLocationResolver.cs:38]: Is there a way to not have to remember to add/sub 1 from the line #s? I feel like this could results in errors or in the tests or something.

File: src/Analyzer.JsonRuleEngine/Expressions/Expression.cs [hunk 21]
```diff
@@ -42,10 +42,10 @@ internal Expression(ExpressionCommonProperties commonProperties)
         /// Executes this <see cref="Expression"/> against a template.
         /// </summary>
         /// <param name="jsonScope">The specific scope to evaluate.</param>
-        /// <param name="jsonLineNumberResolver">An <see cref="ILineNumberResolver"/> to
+        /// <param name="jsonLineNumberResolver">An <see cref="ILocationResolver"/> to
```
Comment (anaismiller) [inline src/Analyzer.JsonRuleEngine/Expressions/Expression.cs:45]: and while updating the comments, also update the Readmes (not sure if it is there). 
And the Readme should say there is Bicep too in the tool.

File: src/Analyzer.Types/TemplateContext.cs [hunk 22]
```diff
@@ -20,6 +20,16 @@ public class TemplateContext
         /// The template with resolved parameter/variable references and expressions
         /// </summary>
         public JToken ExpandedTemplate { get; set; }
+        
+        /// <summary>
+        /// Whether this template was originally a Bicep file
+        /// </summary>
+        public bool IsBicep { get; set; }
```
Comment (anaismiller) [inline src/Analyzer.Types/TemplateContext.cs:27]: Should there be a default value/is there a default value?

File: src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs [hunk 23]
```diff
@@ -116,11 +116,29 @@ public void AnalyzeTemplate_TemplateIsNull_ThrowArgumentNullException()
...
+        public void AnalyzeTemplate_JsonTemplateIsInvalid_ThrowTemplateAnalyzerException()
         {
             templateAnalyzer.AnalyzeTemplate("{}");
         }
 
+        [TestMethod]
+        [ExpectedException(typeof(TemplateAnalyzerException))]
+        public void AnalyzeTemplate_BicepTemplateIsInvalid_ThrowTemplateAnalyzerException()
+        {
+            var invalidBicep = "param location string = badstring";
```
Comment (anaismiller) [inline src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs:128]:
```suggestion
            var invalidBicep = "param location string = badString";
```

File: src/Analyzer.BicepProcessor/BicepLocationResolver.cs (hunk 20 above)
Comment (nonik0) [reply src/Analyzer.BicepProcessor/BicepLocationResolver.cs:45]: leaving as is per offline discussion

File: src/Analyzer.BicepProcessor/BicepLocationResolver.cs (hunk 19 above)
Comment (nonik0) [reply src/Analyzer.BicepProcessor/BicepLocationResolver.cs]: leaving as is per offline discussion

File: src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+    /// <summary>
+    /// Contains functionality to process Bicep templates.
+    /// </summary>
+    public class BicepTemplateProcessor
+    {
+        private static readonly IConfigurationManager configurationManager = new ConfigurationManager(new FileSystem());
+        private static readonly IFileResolver fileResolver = new FileResolver();
+        private static readonly IFeatureProvider featureProvider = new FeatureProvider();
+
+        private static readonly EmitterSettings emitterSettings = new(featureProvider);
```
[the remark this answers is not in the document]
Comment (nonik0) [reply src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs:33]: This is actually just a newer shorthand syntax for constructors I first saw working on bicep: `Type t = new Type(arg1)` is redundant as the type is specified twice, so before you could do `var t = new Type(arg1)` where the type is inferred for the var but now you can also do `Type t = new(arg1)` where the constructor is inferred from the var type. 😎

I could technically do it for the other private members but I am instead using the generic interface which prevents me from doing this shorthand.

File: src/Analyzer.JsonRuleEngine.UnitTests/ExpressionTests.cs [hunk 25]
```diff
@@ -173,7 +173,7 @@ public void EvaluateInternal_HasWhereCondition_LineNumberResolverNotPassed()
             };
 
             // Evaluate scope - line number resolver should not be passed into Where condition
-            mockExpression.Evaluate(mockPathResolver.Object, new Mock<ILineNumberResolver>().Object);
+            mockExpression.Evaluate(mockPathResolver.Object, new Mock<ILocationResolver>().Object);
```
[the remark this answers is not in the document]
Comment (nonik0) [reply src/Analyzer.JsonRuleEngine.UnitTests/ExpressionTests.cs:176]: I reverted to ILineNumberResolver--these updates can be delayed to when they are needed and the actual interface needs to change.

File: src/Analyzer.Types/TemplateContext.cs (hunk 22 above)
Comment (nonik0) [reply src/Analyzer.Types/TemplateContext.cs:27]: I think the default value would be false for a bool. There's only one location in Analyzer.Core where TemplateContext is created that will always directly set this value as well.

File: src/Analyzer.Core/TemplateAnalyzer.cs [hunk 26]
```diff
@@ -73,6 +79,21 @@ public IEnumerable<IEvaluation> AnalyzeTemplate(string template, string paramete
         {
             if (template == null) throw new ArgumentNullException(nameof(template));
 
+            // if the template is bicep, convert to JSON and get source map
+            var isBicep = templateFilePath != null && templateFilePath.ToLower().EndsWith(".bicep", StringComparison.OrdinalIgnoreCase);
```
[the remark this answers is not in the document]
Comment (nonik0) [reply src/Analyzer.Core/TemplateAnalyzer.cs:83]: leaving as is per offline discussion

File: src/Analyzer.JsonRuleEngine/Expressions/Expression.cs (hunk 21 above)
Comment (nonik0) [reply src/Analyzer.JsonRuleEngine/Expressions/Expression.cs:45]: READMEs are updated

File: README.md [hunk 27]
```diff
@@ -9,6 +9,8 @@
 
 The Template BPA provides a simple and extensible solution to improve the security of your Azure resources before deployment and ensures your ARM templates follow best practices. The Template BPA is designed to be customizable - users can write their own checks and/or enforce only the checks that are relevant for them.
 
+The Template BPA is also capable of scanning newer [Bicep template](https://docs.microsoft.com/en-us/azure/azure-resource-manager/bicep/) files.
```
Comment (anaismiller) [inline README.md:12]:
```suggestion
The Template BPA is also capable of scanning newer [Bicep template](https://docs.microsoft.com/azure/azure-resource-manager/bicep/) files.
```
```suggestion
The Template BPA is also capable of scanning newer [Bicep template](https://docs.microsoft.com/en-us/azure/azure-resource-manager/bicep/) files.
```

File: README.md [hunk 28]
```diff
@@ -9,6 +9,8 @@
 
 The Template BPA provides a simple and extensible solution to improve the security of your Azure resources before deployment and ensures your ARM templates follow best practices. The Template BPA is designed to be customizable - users can write their own checks and/or enforce only the checks that are relevant for them.
```
Comment (anaismiller) [inline README.md:10]: Why not make it ARM and Bicep templates?

Comment (anaismiller) [reply README.md:10]: In the intro paragraph

File: README.md (hunk 27 above)
Comment (anaismiller) [reply README.md:12]: should there be a link to the arm templates too?

**Review by anaismiller — APPROVED** [review]
minor comments needed to be resolved, approving

File: README.md (hunk 28 above)
Comment (nonik0) [reply README.md:10]: done
